src/boot/chid.h

```diff
@@ -45,19 +55,15 @@ assert_cc(offsetof(Device, descriptor) == 0);
 assert_cc(offsetof(Device, chid) == 4);
 assert_cc(offsetof(Device, devicetree.name_offset) == 20);
 assert_cc(offsetof(Device, devicetree.compatible_offset) == 24);
+assert_cc(offsetof(Device, fw.name_offset) == 20);
+assert_cc(offsetof(Device, fw.compatible_offset) == 24);
 assert_cc(sizeof(Device) == 28);
 
 static inline const char* device_get_name(const void *base, const Device *device) {
-        if (device->descriptor != DEVICE_DESCRIPTOR_DEVICETREE)
-                return NULL;
-
```

Comment:
This is wrong, `device->devicetree` fields may only be accessed for `DEVICE_TYPE_DEVICETREE`. 

I understand why you did this, but this isn't correct, unfortunately. You'll probably have to implement something like https://github.com/systemd/systemd/pull/35281#discussion_r1864938466. `test-chid-match.c` needs to be updated accordingly to pass the expected device type (and maybe also check that it was returned).

Reply:
Not resolved. `device_get_{name,compatible}` will access `.devicetree` fields even for firmware `Device`.

Reply:
Can you please explain how ? It checks if the device descriptor matches that of the passed descriptor

Reply:
The check is absolutely useless.
You're literally passing `dev->descriptor` in the test and it gets compared with itself. For `Device 4` `chid_match` will return device with `DEVICE_DESCRIPTOR_FW` and for those, it's incorrect to access `dev->devicetree`. This of course works because the structure layout is the same, but you shouldn't rely on that.